### ISSUE TITLE
Return value from docker container is wrong with unhandled exceptions and signals

### ISSUE THREAD

**Issue by jrnos** [docker/docker#9351]
The return value from docker run command is wrong when the executed application exits with an unhandled exception or due to a signal. This is reproducable easily with the sources shown below and basically whatever linux container (I have tried with Ubuntu 14.04 and CentOS 6.6).

```c++
#include <cstdlib>
#include <iostream>
#include <stdexcept>
#include <signal.h>
  
using namespace std;
 
int main(int argc, char** argv) {
    if (argc == 1) {
        cout << "Usage: dk mode\n"
             << "Mode can be:\n"
             << "  1 - return with value 1\n"
             << "  2 - Unhandled runtime_error\n"
             << "  3 - Segmentation fault\n"
             << " Other values - return with value 0" << endl;

        return 0;
    }

    switch (atoi(argv[1])) {
        case 1:
            return 1;
        case 2:
            throw runtime_error("Got ya");
        case 3:
            raise(SIGSEGV);
            cout << "Signal raised (if you see this, the signal didn't have immediate effect)" << endl;
            return 234;
        default:
            return 0;
    }
}
```

An example from CentOS 6.6 based distro and CentOS 6.6 container. The used docker version is 1.3.1, build c78088f/1.3.1 and the execution driver is native-0.2. Same results are seen with Ubuntu 14.04 host & containers where docker version is 1.0.1, build 990021a and the execution driver the same native-0.2.

If the test program is executed outside docker container i.e. directly in the host:
```
> ./a.out 0; echo $?
0
> ./a.out 1; echo $?
1
> ./a.out 2; echo $?
terminate called after throwing an instance of 'std::runtime_error'
  what():  Got ya
Aborted
134
> ./a.out 3; echo $?
Segmentation fault
139
```
The same results are obtained if running the commands inside a bash shell running inside a docker container (i.e. running first "docker run -t -i test123 bash" and then the test problem with different arguments inside the shell).

What happens if running the test executable directly inside docker container:
```
> docker run test123 /test/a.out 0; echo $?
0
> docker run test123 /test/a.out 1; echo $?
1
> docker run test123 /test/a.out 2; echo $?
terminate called after throwing an instance of 'std::runtime_error'
  what():  Got ya
255
docker run test123 /test/a.out 3; echo $?
Signal raised (if you see this, the signal didn't have immediate effect)
234
```
As seen, with the parameter 2 the return value 255 is wrong. With the parameter 3 the signal didn't have immediate effect and so 234 is returned.

Funnily enough, one can use a trick with the bash shell to get around the problem (the double colon and semicolon are needed, without those the return value won't be right):
```
> docker run test123 bash -c ":;/test/a.out 2"; echo $?
terminate called after throwing an instance of 'std::runtime_error'
  what():  Got ya
bash: line 1:     6 Aborted                 /test/a.out 2
134
> docker run test123 bash -c ":;/test/a.out 3"; echo $?
bash: line 1:     6 Segmentation fault      /test/a.out 3
139
```

**Comment by phemmer** [comment]
This appears to be a dup of #3793

Docker PID 1 ignores all signals except for SIGSTOP & SIGKILL. In your case 2, the program is getting a SIGABRT, and when in docker, ignores it. In your case 3, it's a SIGSEGV, and also ignored.
For your program to exit upon receiving these signals, you need to register them to SIG_ERR.

**Comment by icecrime** [comment]
Dup of #3793 indeed (which will be fixed with #26061).
